src/communication/websocket/session.cpp

```diff
@@ -86,14 +100,78 @@ void Session::DoRead() {
       }));
 }
 
-void Session::OnRead(boost::beast::error_code ec, size_t /*bytes_transferred*/) {
+void Session::DoClose() {
+  ws_.async_close(boost::beast::websocket::close_code::normal,
+                  boost::asio::bind_executor(strand_, [shared_this = shared_from_this()](boost::beast::error_code ec) {
+                    shared_this->OnClose(ec);
+                  }));
+}
+
+void Session::OnClose(boost::beast::error_code ec) {
+  if (ec) {
+    return LogError(ec, "close");
+  }
+  connected_.store(false, std::memory_order_relaxed);
+}
+
+utils::BasicResult<std::string> Session::Authorize(const nlohmann::json &creds) {
+  if (!auth_.Authenticate(creds.at("username").get<std::string>(), creds.at("password").get<std::string>())) {
+    return {"Authentication failed!"};
+  }
+#ifndef MG_ENTERPRISE
+  if (auth_.UserHasPermission(creds.at("username").get<std::string>(), auth::Permission::WEBSOCKET)) {
+    return {"Authorization failed!"};
+  }
+#endif
+  return {};
+}
+
+void Session::OnRead(const boost::beast::error_code ec, const size_t /*bytes_transferred*/) {
   if (ec == boost::beast::websocket::error::closed) {
     messages_.clear();
     connected_.store(false, std::memory_order_relaxed);
     return;
   }
 
-  buffer_.consume(buffer_.size());
+  if (!authenticated_ && auth_.HasAnyUsers()) {
```

Comment:
```suggestion
  if (auth_.HasAnyUsers() && !authenticated_) {
```
I think it's better to do the simpler operation first because of short circuting

Reply:
Isn't checking `authenticated_` the simpler, as it is just a simple bool member?

Reply:
That is right, I thought it was a call to Authenticate for some reason.... Sorry @jbajic